snowfakery/output_streams.py

```diff
@@ -194,51 +215,42 @@ class JSONOutputStream(OutputStream):
 
     def __init__(self, file):
         assert file
-        self.file = file
+        super().__init__(file)
         self.first_row = True
 
     def write_single_row(self, tablename: str, row: Dict) -> None:
         if self.first_row:
-            self.file.write("[")
+            self.stream.write("[")
             self.first_row = False
         else:
-            self.file.write(",\n")
+            self.stream.write(",\n")
         values = {"_table": tablename, **row}
-        self.file.write(json.dumps(values))
+        self.stream.write(json.dumps(values))
 
     def close(self) -> None:
-        self.file.write("]\n")
+        self.stream.write("]\n")
+        super().close()
 
 
 class SqlOutputStream(OutputStream):
     mappings = None
+    should_close_session = False
 
     def __init__(self, engine: Engine, mappings: Optional[Dict]):
         self.buffered_rows = defaultdict(list)
-        self.engine = None
         self.table_info = {}
         self.mappings = mappings
         self.engine = engine
-        self._init_db()
+        self.session = create_session(bind=self.engine, autocommit=False)
+        self.metadata = MetaData(bind=self.engine)
+        self.base = automap_base(bind=engine, metadata=self.metadata)
 
     @classmethod
     def from_url(cls, db_url: str, mappings: Optional[Dict] = None):
-        self = cls(create_engine(db_url), mappings)
-        return self
-
-    @classmethod
-    def from_connection(
-        cls, session, engine: Engine, base, mappings: Optional[Dict] = None
-    ):
```

Comment:
from_connection is currently used in CumulusCI; I imagine you're intending to make a new release of snowfakery and then update https://github.com/SFDO-Tooling/CumulusCI/pull/1659?